system/src/iosMain/com.splendo.kaluga/system/network/NetworkManager.kt

```diff
@@ -0,0 +1,138 @@
+    See the License for the specific language governing permissions and
+    limitations under the License.
+
+ */
+
+package com.splendo.kaluga.system.network
+
+import co.touchlab.stately.concurrency.AtomicReference
+import com.splendo.kaluga.logging.debug
+import kotlinx.cinterop.COpaquePointer
+import kotlinx.cinterop.StableRef
+import kotlinx.cinterop.alloc
+import kotlinx.cinterop.asStableRef
+import kotlinx.cinterop.nativeHeap
+import kotlinx.cinterop.ptr
+import kotlinx.cinterop.staticCFunction
+import platform.SystemConfiguration.SCNetworkReachabilityCallBack
+import platform.SystemConfiguration.SCNetworkReachabilityContext
+import platform.SystemConfiguration.SCNetworkReachabilityCreateWithName
+import platform.SystemConfiguration.SCNetworkReachabilityFlags
+import platform.SystemConfiguration.SCNetworkReachabilityFlagsVar
+import platform.SystemConfiguration.SCNetworkReachabilityGetFlags
+import platform.SystemConfiguration.SCNetworkReachabilityRef
+import platform.SystemConfiguration.SCNetworkReachabilitySetCallback
+import platform.SystemConfiguration.SCNetworkReachabilitySetDispatchQueue
+import platform.SystemConfiguration.kSCNetworkReachabilityFlagsIsWWAN
+import platform.SystemConfiguration.kSCNetworkReachabilityFlagsReachable
+import platform.darwin.dispatch_get_main_queue
+
+actual class NetworkManager actual constructor(
+    networkStateRepo: NetworkStateRepo,
+    context: Any?
+) : BaseNetworkManager(networkStateRepo) {
+
+    private val reachability = SCNetworkReachabilityCreateWithName(null, "www.appleiphonecell.com")
+
+    private var _isListening = AtomicReference(false)
+    private var isListening: Boolean
+        get() = _isListening.get()
+        set(value) = _isListening.set(value)
+
+    private var _isNetworkEnabled = AtomicReference(false)
+    private var isNetworkEnabled: Boolean
+        get() = _isNetworkEnabled.get()
+        set(value) = _isNetworkEnabled.set(value)
+
+    private val onNetworkStateChanged: SCNetworkReachabilityCallBack = staticCFunction { ref: SCNetworkReachabilityRef?, flags: SCNetworkReachabilityFlags, info: COpaquePointer? ->
+        if (info == null) {
+            debug { "DEBUG: info is null, returning." }
+            return@staticCFunction
+        }
+        val networkManager = info.asStableRef<NetworkManager>().get()
+        networkManager.checkReachability(networkManager, flags)
+    }
+
+    override suspend fun isNetworkEnabled(): Boolean = isNetworkEnabled
+
+    override suspend fun startMonitoringNetwork() {
+        // val currentVersion = UIDevice.currentDevice.systemVersion.split(".")[0].toInt()
+        // if (currentVersion >= 12) {
+        //     println("current ios version is > 12 $currentVersion")
+        //     // configureNwPath()
+        // } else {
+        //     println("current ios version is < 12 $currentVersion")
+        //     configureSCNetworkReachability()
+        // }
```

Comment:
Is there a way to get the version of the device in a safer way?

Reply:
I guess this: https://developer.apple.com/documentation/foundation/nsprocessinfo/1414876-isoperatingsystematleastversion

Reply:
There is a util in Kaluga that @Daeda88 made I think

Reply:
https://github.com/splendo/kaluga/blob/f407b6c6d3b120e99759b95a6ff412cdf2956372/base/src/iosMain/kotlin/utils/IOSVersion.kt#L25